src/Microsoft.ML.Tokenizers/Model/BPE.cs

```diff
@@ -327,80 +317,124 @@ internal string CharToString(char c)
             return s;
         }
 
-        internal Word MergeWord(string w)
+        internal Word MergeWord(ReadOnlySpan<char> w)
         {
             Word word = Word.WithCapacity(w.Length);
             (int Id, int Len)? unk = null;
             int i = 0;
 
-            while (i < w.Length)
-            {
-                int length;
-                string s;
-
-                if (Char.IsHighSurrogate(w[i]) && i < w.Length - 1 && Char.IsLowSurrogate(w[i + 1]))
-                {
-                    length = 2;
-                    s = w.Substring(i, length);
-                }
-                else
-                {
-                    length = 1;
-                    s = CharToString(w[i]);
-                }
+            char[]? buffer = null;
 
-                // Add the `continuing_subword_prefix` if relevant
-                if (i > 0 && ContinuingSubwordPrefix is not null)
+            try
+            {
+                while (i < w.Length)
                 {
-                    s = $"{ContinuingSubwordPrefix}{s}";
-                }
+                    int length;
+                    ReadOnlySpan<char> s;
 
-                // Add the `end_of_word_suffix` if relevant
-                if (i + length >= w.Length && EndOfWordSuffix is not null)
-                {
-                    s = $"{s}{EndOfWordSuffix}";
-                }
+                    if (Char.IsHighSurrogate(w[i]) && i < w.Length - 1 && Char.IsLowSurrogate(w[i + 1]))
+                    {
+                        length = 2;
+                        s = w.Slice(i, 2);
+                    }
+                    else
+                    {
+                        length = 1;
+                        s = w.Slice(i, 1);
+                    }
 
-                if (_vocab.TryGetValue(s, out int id))
-                {
-                    if (unk.HasValue)
+                    // Add the `continuing_subword_prefix` if relevant
+                    if (i > 0 && ContinuingSubwordPrefix is not null)
                     {
-                        word.Add(unk.Value.Id, unk.Value.Len);
-                        unk = null;
+                        if (buffer is null)
+                        {
+                            // 60 should be enough for most cases
+                            buffer = ArrayPool<char>.Shared.Rent(60);
```

Comment:
If this is only ever renting 60 chars, could you just stackalloc it at the beginning of the method? You could do so for more than 60, e.g. 256, too.

Reply:
Actually I tried to use stackalloc first but I got the following error. I couldn't figure out how I can work around that even the buffer usage scoped to the declared method. 

```
error CS8352: Cannot use variable 'buffer' in this context because it may expose referenced variables outside of their declaration scope
```

Any thoughts if there is a work around for this error?

Reply:
never mind, I used scoped keyword and worked fine.